src/python/foglamp/device/ingest.py

```diff
@@ -0,0 +1,180 @@
+
+    @classmethod
+    async def stop(cls):
+        if cls._stop or cls._write_statistics_loop_task is None:
+            return
+
+        cls._stop = True
+
+        if cls._sleep_task is not None:
+            cls._sleep_task.cancel()
+            cls._sleep_task = None
+
+        await cls._write_statistics_loop_task
+        cls._write_statistics_loop_task = None
+
+    @classmethod
+    def increment_discarded_messages(cls):
+        cls._num_discarded_readings += 1
+
+    @classmethod
+    async def _write_statistics_loop(cls):
+        """Periodically commits collected readings statistics"""
+        _LOGGER.info("Ingest statistics writer started")
+
+        while not cls._stop:
+            # stop() calls _sleep_task.cancel().
+            # Tracking _sleep_task separately is cleaner than canceling
+            # this entire coroutine because allowing database activity to be
+            # interrupted will result in strange behavior.
+            cls._sleep_task = asyncio.ensure_future(
+                                asyncio.sleep(_STATISTICS_WRITE_FREQUENCY_SECONDS))
+
+            try:
+                await cls._sleep_task
+            except asyncio.CancelledError:
+                pass
+
+            cls._sleep_task = None
+
+            try:
+                # TODO Move READINGS and DISCARDED to globals
+                await statistics.update_statistics_value('READINGS', cls._num_readings)
+                cls._num_readings = 0
+
+                await statistics.update_statistics_value('DISCARDED', cls._num_discarded_readings)
+                cls._num_discarded_readings = 0
+            # TODO catch real exception
+            except Exception:
+                _LOGGER.exception("An error occurred while writing readings statistics")
+
+        _LOGGER.info("Ingest statistics writer stopped")
+
+    @classmethod
+    async def add_readings(cls, data: dict)->None:
+        """Sends asset readings to storage layer
+
+        Args:
+            data:
+            {
+                "timestamp": "2017-01-02T01:02:03.23232Z-05:00",
+                "asset": "pump1",
+                "readings": {
+                    "velocity": "500",
+                    "temperature": {
+                        "value": "32",
+                        "unit": "kelvin"
+                    }
+                }
+            }
+
+        Raises KeyError: data is missing a required field
+        Raises IOError: some type of failure occurred
+        Raises TypeError: bad data provided
+        """
+
+        # TODO: The data format is documented
+        # at https://docs.google.com/document/d/1rJXlOqCGomPKEKx2ReoofZTXQt9dtDiW_BHU7FYsj-k/edit#
+        # and will be moved to a .rst file
+
+        cls._num_readings += 1
+
+        # Required keys in the data
+        try:
+            asset = data['asset']
+            timestamp = data['timestamp']
+        except KeyError:
```

Comment:
TypeError should definitely be thrown out of this method. That's the general way all Python code should handle use cases where bad data is passed in.